mypyc/irbuild/specialize.py

```diff
@@ -248,7 +248,11 @@ def gen_inner_stmts() -> None:
 
 @specialize_function('builtins.isinstance')
 def translate_isinstance(builder: IRBuilder, expr: CallExpr, callee: RefExpr) -> Optional[Value]:
-    # Special case builtins.isinstance
+    # Special case for builtins.isinstance
+    # Prevent coercions on the thing we are checking the instance of - there is no need to coerce
+    # something to a new type before checking what type it is, and the coercion could lead to bugs.
+    builder.types[expr.args[0]] = AnyType(TypeOfAny.from_error)
```

Comment:
This could be moved to within the if statement, so that if somehow there aren't any arguments, we won't get an `IndexError`.